docs/GDPs/gdp-duck-storage.rst

```diff
@@ -0,0 +1,456 @@
+the `ExplicitlySyncedGPUStorage` could not be tracked reliably, resulting in validation errors which are hard to
+debug and avoid.
+
+Nevertheless, the storages are currently implemented as ndarray subclasses. This was necessary
+so the storages could be used in some third party frameworks. Since
+then, these frameworks have extended their compatibility to the interface that are
+specified in the :emphasis:`Numpy Enhancement Proposal`
+`NEP18 <https://numpy.org/neps/nep-0018-array-function-protocol.html>`_.
+
+Implementing the interface introduced in the NEP18 allows GT4Py to retain full control
+over the copying behavior, while allowing the use of the mentioned frameworks.
+
+Besides this substantial change, we propose to take the opportunity of this development
+to also improve the interaction with existing codes by allowing the initialization of
+storages from external buffers without copying. To use this feature, some additional
+information about the provided buffers needs to be specified to both the :code:`__init__`
+method of the storages as well to the stencils at compile time.
+
+Despite the move away from NumPy as a base class, the storage should integrate with the
+scientific python ecosystem as seamlessly as possible and mimic a NumPy ndarray, which is an established standard,
+by means of `duck typing <https://en.wikipedia.org/wiki/Duck_typing>`_, to the extent that this is feasible.
+
+Backward Compatibility
+----------------------
+
+The implementation of this GDP breaks the integration with external libraries which require a NumPy ndarray subclass.
+Further, we propose some API changes like renaming or repurposing of keyword arguments and attributes.
+This will break all existing user codes. However, all existing GT4Py functionality will remain or be extended. Updating
+codebases to the new interface amounts mostly to updating attribute and keyword argument names.
+
+
+Detailed Description
+--------------------
+
+Supported Functionality
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Ideally, storages would behave in the same way as ndarrays. We limit this if it is not possible to maintain the
+consistency of the underlying structured memory. We chose to internally use NumPy and CuPy ndarrays
+to store CPU and GPU buffers, respectively. We also implement some functionality like mathematical operations by
+forwarding to those libraries. Therefore, support for dtypes and other functionality is restricted to the common
+denominator of CuPy and NumPy.
+
+
+Storage Constructors
+^^^^^^^^^^^^^^^^^^^^
+
+There will be 6 functions that should be used to initialize the storages. These are:
+
+:code:`empty`
+   Used to allocate a storage with uninitialized (undefined) values.
+:code:`zeros`
+   Used to allocate a storage with values initialized to 0.
+:code:`ones`
+   Used to allocate a storage with values initialized to 1.
+:code:`full`
+   Used to allocate a storage with values initialized to a given scalar value.
+:code:`asstorage`
```

Comment:
I think it is "explicit" enough given that it is the equivalent to numpy's array/asarray. we could of course just call everything everywhere ndarray instead of storage, and then call these routines array/asarray, like numpy